src/mesh/BoundingBox.cpp

```diff
@@ -108,58 +105,67 @@ int BoundingBox::getDimension() const
 
 const std::vector<double> BoundingBox::dataVector() const
 {
-  return _bounds;
+  std::vector<double> bounds;
+  bounds.insert(bounds.end(), {_boundMin[0], _boundMax[0], _boundMin[1], _boundMax[1]});
+  if (_dimensions == 3) {
+    bounds.insert(bounds.end(), {_boundMin[2], _boundMax[2]});
+  }
+  return bounds;
 }
 
 void BoundingBox::expandBy(const BoundingBox &otherBB)
 {
+  PRECICE_ASSERT(_dimensions == otherBB.getDimension(), "Other BoundingBox with different dimensions than this bounding box cannot be used to expand it.");
   for (int d = 0; d < _dimensions; d++) {
-    _bounds[2 * d]     = std::min(_bounds[2 * d], otherBB._bounds[2 * d]);
-    _bounds[2 * d + 1] = std::max(_bounds[2 * d + 1], otherBB._bounds[2 * d + 1]);
+    _boundMin[d] = std::min(_boundMin[d], otherBB._boundMin[d]);
+    _boundMax[d] = std::max(_boundMax[d], otherBB._boundMax[d]);
   }
 }
 
 void BoundingBox::expandBy(const Vertex &vertices)
 {
-  PRECICE_ASSERT(_dimensions == vertices.getDimensions(), "Vertex with different dimensions than bounding box cannot be used to expand bounding box");
+  PRECICE_ASSERT(_dimensions == vertices.getDimensions(), "Vertex with different dimensions than this bounding box cannot be used to expand it.");
   const auto coords = vertices.rawCoords();
   for (int d = 0; d < _dimensions; ++d) {
-    _bounds.at(2 * d)     = std::min(coords[d], _bounds.at(2 * d));
-    _bounds.at(2 * d + 1) = std::max(coords[d], _bounds.at(2 * d + 1));
+    _boundMin[d] = std::min(coords[d], _boundMin[d]);
+    _boundMax[d] = std::max(coords[d], _boundMax[d]);
   }
 }
 
 void BoundingBox::expandBy(double value)
 {
-  if (empty())
+  // Adding or subtracting a constant to the default state will still make an 'illegal' box
+  if (isDefault()) {
     return;
+  }
```

Comment:
Could you add this behavior to the function description? I'm still a bit unhappy that this function overloads with the previous two, but still behaves differently (in this way).

Also, you can just wrap the code into the if block and remove the return, as it is a void function.

Reply:
We add this behavior inside this function as a comment.
```cpp
void BoundingBox::expandBy(double value)
{
  // BoundingBox should not be default state; otherwise, it will make ReceivedPartitionTest fail because of incorrect connectionMapSize
  if (!isDefault()) {
    for (int d = 0; d < _dimensions; d++) {
      _boundMin[d] -= value;
      _boundMax[d] += value;
    }
  }
}
```
Error message of `ReceivedPartitionTest` fail because of incorrect `connectionMapSize`:
```
gney@Hsiao:~/projects/advprog_bp/precice/build$ ctest --rerun-failed --output-on-failure
Test project /home/gney/projects/advprog_bp/precice/build
    Start 10: precice.partition
1/1 Test #10: precice.partition ................***Timeout  20.02 sec
[2023-01-29 18:42:33.002000] [0x00007f04d711b780] [debug]   Initialize MPI
[2023-01-29 18:42:33.001894] [0x00007f98e492a780] [debug]   Initialize MPI
[2023-01-29 18:42:33.002640] [0x00007f0e91707780] [debug]   Initialize MPI
[2023-01-29 18:42:33.005487] [0x00007f9353450780] [debug]   Initialize MPI
This test suite runs on rank 1 of 4
This test suite runs on rank 2 of 4
This test suite runs on rank 0 of 4
This test suite runs on rank 3 of 4
Running 20 test cases...
Running 20 test cases...
Running 20 test cases...
Setup up logging
Setup up logging
Setup up logging
Running 20 test cases...
Setup up logging
18:42:33.586897|Fluid|0|partition::ReceivedPartition|l891|createOwnerInformation|WARNING: 2 of 6 vertices of mesh SolidzMesh have been filtered out since they have no influence on the mapping.
18:42:33.638039|Fluid|0|partition::ReceivedPartition|l891|createOwnerInformation|WARNING: 1 of 3 vertices of mesh NastinMesh have been filtered out since they have no influence on the mapping.
/home/gney/projects/advprog_bp/precice/src/partition/tests/ReceivedPartitionTest.cpp(949): fatal error: in "PartitionTests/ReceivedPartitionTests/TestCompareBoundingBoxes2D": critical check connectionMapSize == 2 has failed [3 != 2]
Failure occurred in a following context:
    Test context represents "SOLIDZ" and runs on rank 0 out of 1. Initialized: { Events}
/home/gney/projects/advprog_bp/precice/src/partition/tests/ReceivedPartitionTest.cpp(1017): error: in "PartitionTests/ReceivedPartitionTests/TestCompareBoundingBoxes3D": check connectionMapSize == 2 has failed [3 != 2]
Failure occurred in a following context:
    Test context represents "SOLIDZ" and runs on rank 0 out of 1. Initialized: { Events}
18:42:33.764675||3|partition::ReceivedPartition|l745|createOwnerInformation|WARNING: 1 of 6 vertices of mesh mesh have been filtered out since they have no influence on the mapping.
18:42:33.764673||0|partition::ReceivedPartition|l745|createOwnerInformation|WARNING: 1 of 6 vertices of mesh mesh have been filtered out since they have no influence on the mapping.
18:42:33.764959||1|partition::ReceivedPartition|l745|createOwnerInformation|WARNING: 1 of 6 vertices of mesh mesh have been filtered out since they have no influence on the mapping.
18:42:33.771773||1|partition::ReceivedPartition|l745|createOwnerInformation|WARNING: 2 of 12 vertices of mesh mesh have been filtered out since they have no influence on the mapping.
18:42:33.771742||3|partition::ReceivedPartition|l745|createOwnerInformation|WARNING: 2 of 12 vertices of mesh mesh have been filtered out since they have no influence on the mapping.
18:42:33.771858||2|partition::ReceivedPartition|l745|createOwnerInformation|WARNING: 2 of 12 vertices of mesh mesh have been filtered out since they have no influence on the mapping.


0% tests passed, 1 tests failed out of 1

Total Test time (real) =  20.02 sec

The following tests FAILED:
         10 - precice.partition (Timeout)
Errors while running CTest
```

Reply:
> We (@jeremiahbintang) add this behavior inside this function as a comment.

There is no real value in adding the docs to the function implementation. We generate code documentation based on doxygen, which usually looks at the function declaration (in the header file). Please at the documentation there such that we don't have to look for the implementation each time.

Reply:
We (@jeremiahbintang ) have added the description in the hpp file.
```cpp
/// Expand bounding box using double value
/// Using this method, BoundingBox should not be in the default state; otherwise, it will make ReceivedPartitionTest fail because of incorrect connectionMapSize
  void expandBy(double value);
```

Reply:
> Expand bounding box using double value

To which direction / how?

> otherwise, it will make ReceivedPartitionTest fail because of incorrect connectionMapSize

I am not sure if referring to the test is a good idea here. It is not that the function is made for the test, but the test for the function.

But I think you probably just don't have enough context to write a more detailed comment about the motivation (neither do I :grimacing: ).

Reply:
1. 'expandBy(double value)' Expand bounding box using a double value in all dimensions.
2. Yes, we still don't know why it fails the test😅 The reason why we wrote it down is to record the behavior clearly for other developers' reference.